File: unified-codes/libs/data/src/lib/types/Property.ts
export interface IProperty {
  type: string;
  value: string;
  properties?: IProperty[];
}

export class Property {
  private _type: string;
  private _value: string;
  private _properties?: Property[];

  constructor(property: IProperty) {
    this._type = property.type;
    this._value = property.value;
    this._properties = property.properties?.map((property: IProperty) => new Property(property));
  }

  get type(): string {
    return this._type;
  }

  get value(): string {
    return this._value;
  }

  get properties(): Property[] | undefined {
    return this._properties;
  }
}

export default IProperty;

File: backend/graphql_v1/universal_codes/src/types/property.rs
use async_graphql::*;
use dgraph::Property;

#[derive(Clone, Debug)]
pub struct PropertiesType {
    pub code: String,
    pub key: String,
    pub value: String,
}

#[Object]
impl PropertiesType {
    pub async fn code(&self) -> &str {
        &self.code
    }

    pub async fn r#type(&self) -> &str {
        &self.key
    }

    pub async fn value(&self) -> &str {
        &self.value
    }
}

impl PropertiesType {
    pub fn from_domain(properties: Vec<Property>) -> Vec<PropertiesType> {
        properties
            .into_iter()
            .map(|p| PropertiesType {
                code: p.code,
                key: p.key,
                value: p.value,
            })
            .collect()
    }
}
